### PR TITLE
[Xaml] better error reporting

### LINKED ISSUE #5095: No XAML parsing error with an erroneous character in XAML
### Description

An erroneous character in XAML, if it is inside the root element but outside any other elements, does not cause a parse exception at compile time with Xaml compilation enabled. However in a certain circumstance it can cause the Xaml to not render or a seemingly unrelated error is presented

### Steps to Reproduce

1. Create a template Xamarin.Forms project.
2. Open MainPage.xaml file. 
3. Add the character "8" (or any other character) right before the closing ContentPage tag:
 ```
<ContentPage ...>
   <StackLayout>
      <Label ... />
   </StackLayout>
    8
</ContentPage>
```
4. Run the project

### Expected Behavior
A Xaml Parse error at compile time (or runtime if Xaml compilation is turned off)

### Actual Behavior
No error at compile time but the label is not rendered and the text is not shown. 

### Notes
If instead the erroneous character is placed after the opening ContentPage element, e.g.:

```
<ContentPage ...>
   8
   <StackLayout>
      <Label ... />
   </StackLayout>
</ContentPage>
```
There is no Xaml parse exception either, but the Xaml is rendered as expected, i.e. the label text shows.

If instead the erroneous character is placed inside the StackLayout, but outside any elements, e.g.:

```
<ContentPage ...>
   <StackLayout>
     8 
      <Label ... />
   </StackLayout>
</ContentPage>
```
Then I get a build failure and a compile time error, but it does not say anything about xaml parsing, instead the error is:
```
.../MainPage.xaml: Error: Value cannot be null.
Parameter name: key (App1)
```

Also if Xaml Compilation is turned off, reasonable errors are hit at runtime, so this is just an issue when Xaml compilation is in use. 

### Basic Information

- Version with issue: 3.3.0, 3.4.0, and 4.0.0-pre3
- Last known good version: unknown
- IDE: VS for Mac Version 7.7.3 (build 43)
- Platform Target Frameworks: iOS
  - iOS:  12.1
  - Android: Did not test Android
  - UWP:  Did not test

Originally reported in Forum thread: https://stackoverflow.com/questions/54448640/what-setting-provides-me-with-a-warning-error-for-the-following-xaml

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

Report error on invalid value nodes

### Issues Resolved ### 

- fixes #5095

### API Changes ###

 
 None

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by StephaneDelcroix** [comment]
this is targeted to master as it could cause some xaml files to stop compiling

File: Xamarin.Forms.Build.Tasks/NodeILExtensions.cs [hunk 1]
```diff
@@ -54,8 +54,47 @@ public static bool CanConvertValue(this ValueNode node, ILContext context, TypeR
 				return toType.InheritsFromOrImplements(targetTypeRef);
 			}
 
-			///No reason to return false
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Build.Tasks/NodeILExtensions.cs]: Oh dude, I was sooo wrong

**Comment by StephaneDelcroix** [comment]
note to self:
>Xamarin.Forms.Xaml.UnitTests\AutomationProperties.xaml(11,4): Error : Position 11:4. No property, bindable property, or event found for 'IsInAccessibleTree', or mismatching type between value and property.
Xamarin.Forms.Xaml.UnitTests\Issues\Bz24910.xaml(7,63): Error : Position 7:63. No property, bindable property, or event found for 'NullableInt', or mismatching type between value and property.
Xamarin.Forms.Xaml.UnitTests\Validation\TypeMismatch.xaml(7,16): Error : Position 7:16. No property, bindable property, or event found for 'Content', or mismatching type between value and property.
Process 'msbuild.exe' exited with code '1'.

File: Xamarin.Forms.Xaml.UnitTests/Validation/TypeMismatch.xaml.cs [hunk 2]
```diff
@@ -21,11 +19,18 @@ public TypeMismatch (bool useCompiledXaml)
...
-			public void ThrowsOnMismatchingType (bool useCompiledXaml)
+			[SetUp] public void Setup() => Device.PlatformServices = new MockPlatformServices();
+			[TearDown] public void TearDown() => Device.PlatformServices = null;
+
+			[Test]
+			public void ThrowsOnMismatchingType ([Values(true, false)]bool useCompiledXaml)
 			{
-				Assert.Throws (new XamlParseExceptionConstraint (7, 16, m => m.StartsWith ("Cannot assign property", StringComparison.Ordinal)), () => new TypeMismatch (useCompiledXaml));
+				if (useCompiledXaml)
+					//MockCompiler.Compile(typeof(TypeMismatch));
```
Comment (paymicro) [inline Xamarin.Forms.Xaml.UnitTests/Validation/TypeMismatch.xaml.cs]: This comment is probably unnecessary.

File: Xamarin.Forms.Build.Tasks/NodeILExtensions.cs [hunk 3]
```diff
@@ -54,8 +54,49 @@ public static bool CanConvertValue(this ValueNode node, ILContext context, TypeR
...
+			if (targetTypeRef.FullName == "System.TimeSpan")
+				return true;
+			if (targetTypeRef.FullName == "System.DateTime")
+				return true;
+			if (targetTypeRef.FullName == "System.String")
+				return true;
+			if (targetTypeRef.FullName == "System.Object")
+				return true;
+			if (targetTypeRef.FullName == "System.Decimal")
+				return true;
```
Comment (paymicro) [inline Xamarin.Forms.Build.Tasks/NodeILExtensions.cs]: Oh. Perhaps to use hashset of full names?
```C#
static HashSet<string> SimpleValueTypes = new HashSet<string> { typeof(decimal).FullName, typeof(bool).FullName, ... }

if (SimpleValueTypes.Contains(targetTypeRef.FullName))
return true;
```
or compare types ```HashSet<Type> { typeof(decimal), typeof(bool), ... }```

Comment (paymicro) [reply Xamarin.Forms.Build.Tasks/NodeILExtensions.cs]: or even easier 
```C#
if (targetTypeRef.IsValueType || targetTypeRef.FullName == "System.Object")
```

**Pull request merged**